stdlib/dynarray.mli

```diff
@@ -0,0 +1,418 @@
+(*     en Automatique.                                                    *)
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+(** Dynamic arrays.
+
+    The {!Array} module provide arrays of fixed length. {!Dynarray}
+    provides array whose length can change over time, by adding or
+    removing elements at the end of the array.
+
+    This is typically used to accumulate elements whose number is not
+    known in advance or changes during computation, while also
+    providing fast access to elements at arbitrary positions.
+
+{[
+    let dynarray_of_list li =
+      let arr = Dynarray.create () in
+      List.iter (fun v -> Dynarray.add_last arr v) li;
+      arr
+]}
+
+    The {!Buffer} module provides similar features, but it is
+    specialized for accumulating characters into a dynamically-resized
+    string.
+
+    The {!Stack} module provides a last-in first-out data structure
+    that can be easily implemented on top of dynamic arrays.
+
+    {b Warning.} In their current implementation, the memory layout
+    of dynamic arrays differs from the one of {!Array}s. See the
+    {{!section:memory_layout} Memory Layout} section for more information.
+
+    @since 5.1
+*)
+
+(** {b Unsynchronized accesses} *)
+
+[@@@alert unsynchronized_accesses
+    "Unsynchronized accesses to dynamic arrays are a programming error."
+]
+
+(**
+   Unsynchronized accesses to a dynamic array may lead to an invalid
+   dynamic array state. Thus, concurrent accesses to dynamic arrays
+   must be synchronized (for instance with a {!Mutex.t}).
+*)
+
+(** {1:dynarrays Dynamic arrays} *)
+
+type 'a t
+(** A dynamic array containing values of type ['a].
+
+    A dynamic array [a] is an array, that is, it provides
+    constant-time [get] and [set] operation on indices between [0] and
+    [Dynarray.length a - 1] included. Its {b length} may change over
+    time by adding or removing elements to the end of the array.
+*)
+
+val create : unit -> 'a t
+(** [create ()] is a new, empty array. *)
+
+val make : int -> 'a -> 'a t
+(** [make n x] makes a array of length [n], filled with [x]. *)
+
+val init : int -> (int -> 'a) -> 'a t
+(** [init n f] is a new array [a] of length [n],
+    such that [get a i] is [f i]. In other words,
+    [a] is the the array whose elements are
+    [f 0; f 1; f 2; ...; f (n - 1)].
+
+    This is similar to {!Array.init}. *)
+
+val get : 'a t -> int -> 'a
+(** [get a i] is the [i]-th element of [a], starting with index [0].
+
+    @raise Invalid_argument if the index is
+      invalid (not in [0 .. length a-1]). *)
```

Comment:
I am adding the following definition in the documentation of ['a t].

> We say that an index into a dynarray [a] is valid if it is in
> [0 .. length a - 1] and invalid otherwise.